proposals/NNNN-isolated-synchronous-deinit.md

```diff
@@ -0,0 +1,419 @@
+    // to cross between otherMaria and self.
+    self.friend = otherMaria.friend
+  }
+
+  deinit {
+    // Used to be a potential data race. Now, deinit is also
+    // isolated on the MainActor, so this code is perfectly 
+    // correct.
+    friend.state += 1
+  }
+}
+
+func example() async {
+  let m1 = await Maria()
+  let m2 = await Maria(sharingFriendOf: m1)
+  doSomething(m1, m2)
+} 
+```
+
+In the case of escaping `self`, the race condition is eliminated but the problem of dangling reference remains. This problem exists for synchronous code as well and is orthogonal to the concurrency features.
+
+```swift
+actor Clicker {
+  var count: Int = 0
+
+  func click(_ times: Int) {
+    for _ in 0..<times {
+      self.count += 1 
+    }
+  }
+
+  deinit {
+    let old = count
+    let moreClicks = 10000
+    
+    Task { await self.click(moreClicks) } // ❌ This WILL keep `self` alive after the `deinit`!
+
+    for _ in 0..<moreClicks {
+        // No data race.
+        // Actor job created by the task is either
+        // not created yet or is waiting in the queue.
+      self.count += 1 
+    }
+
+    assert(count == old + moreClicks) // Always works
+    assert(count == old + 2 * moreClicks) // Always fails
+  }
+}
+```
+
+## Detailed design
+
+### Runtime
+
+This proposal introduces a new runtime function that is used to schedule `deinit`'s code on an executor by wrapping it into a task-less ad hoc job. It does not do any reference counting and can be safely used even with references that were released for the last time but not deallocated yet.
+
+If no switching is needed, then `deinit` is executed immediately on the current thread. Otherwise, the task-less job copies priority and task-local values from the task/thread that released the last strong reference to the object.
+
+```cpp
+using DeinitWorkFunction = SWIFT_CC(swift) void (void *);
+  
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+void swift_task_deinitOnExecutor(void *object, DeinitWorkFunction *work, ExecutorRef newExecutor);
+```
+
+```swift
+@_silgen_name("swift_task_deinitOnExecutor")
+@usableFromInline
+internal func _deinitOnExecutor(_ object: __owned AnyObject,
+                                _ work: @convention(thin) (__owned AnyObject) -> Void,
+                                _ executor: UnownedSerialExecutor)
+```
+
+If `deinit` is isolated, code that normally is emitted into `__deallocating_init` gets emitted into a new entity (`__isolated_deallocating_init`), and `__deallocating_init` is emitted as a thunk that reads the executor (from `self` for actors and from the global actor for GAITs) and calls `swift_task_performOnExecutor` passing `self`, `__isolated_deallocating_init` and the desired executor.
+
+![Last release from background thread](./NNNN-isolated-synchronous-deinit/last-release.png)
+
+![Deinit on main thread](./NNNN-isolated-synchronous-deinit/isolated-deinit.png)
```

Comment:
Do you really think these images are essential for understanding the proposal? Considering no other proposal currently has attached images I think it might be good to avoid this, but if you feel there's not another good way to present the information here I'll discuss with the workgroup how we'd like to store images going forward.

Reply:
Removed